src/ray/object_manager/pull_manager.cc

```diff
@@ -103,31 +108,64 @@ bool PullManager::ActivateNextPullBundleRequest(const Queue &bundles,
     return false;
   }
 
-  RAY_LOG(DEBUG) << "Activating request " << next_request_it->first
-                 << " num bytes being pulled: " << num_bytes_being_pulled_
-                 << " num bytes available: " << num_bytes_available_;
-  // Activate the pull bundle request.
-  for (const auto &ref : next_request_it->second.objects) {
+  // Activate the pull bundle request if possible.
+  {
     absl::MutexLock lock(&active_objects_mu_);
-    auto obj_id = ObjectRefToId(ref);
-    bool start_pull = active_object_pull_requests_.count(obj_id) == 0;
-    active_object_pull_requests_[obj_id].insert(next_request_it->first);
-    if (start_pull) {
-      RAY_LOG(DEBUG) << "Activating pull for object " << obj_id;
-      // This is the first bundle request in the queue to require this object.
-      // Add the size to the number of bytes being pulled.
-      auto it = object_pull_requests_.find(obj_id);
-      RAY_CHECK(it != object_pull_requests_.end());
-      num_bytes_being_pulled_ += it->second.object_size;
-      objects_to_pull->push_back(obj_id);
 
-      ResetRetryTimer(obj_id);
+    // First calculate the bytes we need.
+    size_t bytes_to_pull = 0;
+    for (const auto &ref : next_request_it->second.objects) {
+      auto obj_id = ObjectRefToId(ref);
+      bool needs_pull = active_object_pull_requests_.count(obj_id) == 0;
+      if (needs_pull) {
+        // This is the first bundle request in the queue to require this object.
+        // Add the size to the number of bytes being pulled.
+        auto it = object_pull_requests_.find(obj_id);
+        RAY_CHECK(it != object_pull_requests_.end());
+        bytes_to_pull += it->second.object_size;
+      }
+    }
+
+    // Quota check.
+    if (respect_quota && num_active_bundles_ >= min_active_pulls_ &&
+        (num_bytes_being_pulled_ + bytes_to_pull - pinned_objects_size_ >
+         num_bytes_available_)) {
+      RAY_LOG(DEBUG) << "Bundle would exceed quota: "
+                     << "num_bytes_being_pulled(" << num_bytes_being_pulled_
+                     << ") + "
+                        "bytes_to_pull("
+                     << bytes_to_pull
+                     << ") - "
+                        "pinned_objects_size("
+                     << pinned_objects_size_
+                     << ") > "
+                        "num_bytes_available("
+                     << num_bytes_available_ << ")";
+      return false;
+    }
+
+    RAY_LOG(DEBUG) << "Activating request " << next_request_it->first
```

Comment:
I think this used to be a bit cleaner because it was safe to go over the quota temporarily and reverse it later on during Deactivate. Is it structured this way now because we also have to pin the objects and compute the new pinned_objects_size_?

Reply:
I ended up rewriting this to reduce the log spam during debugging. It also is slightly more efficient due to avoiding the repeated pin/unpin calls now, so I think it's a net win.